k8s/pkg/prefixcollector/collector.go

```diff
@@ -0,0 +1,145 @@
+)
+
+func getExcludedPrefixesChan(clientset *kubernetes.Clientset) (<-chan prefix_pool.PrefixPool, error) {
+	prefixes := getExcludedPrefixesFromEnv()
+
+	// trying to get excludePrefixes from kubeadm-config, if it exists
+	if configMapPrefixes, err := getExcludedPrefixesFromConfigMap(clientset); err == nil {
+		poolCh := make(chan prefix_pool.PrefixPool, 1)
+		pool, err := prefix_pool.NewPrefixPool(append(prefixes, configMapPrefixes...)...)
+		if err != nil {
+			return nil, err
+		}
+		poolCh <- pool
+		return poolCh, nil
+	}
+
+	// seems like we don't have kubeadm-config in cluster, starting monitor client
+	return monitorSubnets(clientset, prefixes...), nil
+}
+
+func getExcludedPrefixesFromEnv() []string {
+	excludedPrefixesEnv, ok := os.LookupEnv(ExcludedPrefixesEnv)
+	if !ok {
+		return []string{}
+	}
+	logrus.Infof("Getting excludedPrefixes from ENV: %v", excludedPrefixesEnv)
+	return strings.Split(excludedPrefixesEnv, ",")
+}
+
+func getExcludedPrefixesFromConfigMap(clientset *kubernetes.Clientset) ([]string, error) {
+	kubeadmConfig, err := clientset.CoreV1().
+		ConfigMaps("kube-system").
+		Get("kubeadm-config", metav1.GetOptions{})
+	if err != nil {
+		logrus.Error(err)
+		return nil, err
+	}
+
+	clusterConfiguration := &v1alpha3.ClusterConfiguration{}
+	err = yaml.NewYAMLOrJSONDecoder(strings.NewReader(kubeadmConfig.Data["ClusterConfiguration"]), 4096).
+		Decode(clusterConfiguration)
+	if err != nil {
+		return nil, err
+	}
+
+	podSubnet := clusterConfiguration.Networking.PodSubnet
+	serviceSubnet := clusterConfiguration.Networking.ServiceSubnet
+
+	if podSubnet == "" {
+		return nil, fmt.Errorf("clusterConfiguration.Networking.PodSubnet is empty")
+	}
+	if serviceSubnet == "" {
+		return nil, fmt.Errorf("clusterConfiguration.Networking.ServiceSubnet is empty")
+	}
+
+	return []string{
+		podSubnet,
+		serviceSubnet,
+	}, nil
+}
+
+func monitorSubnets(clientset *kubernetes.Clientset, additionalPrefixes ...string) <-chan prefix_pool.PrefixPool {
+	logrus.Infof("Start monitoring prefixes to exclude")
+	poolCh := make(chan prefix_pool.PrefixPool)
+
+	go func() {
+		for {
+			errCh := make(chan error, 1)
+			go monitorReservedSubnets(poolCh, errCh, clientset, additionalPrefixes)
+			err := <-errCh
+			logrus.Error(err)
+		}
+	}()
+
+	return poolCh
+}
+
+func monitorReservedSubnets(poolCh chan<- prefix_pool.PrefixPool, errCh chan<- error, clientset *kubernetes.Clientset, additionalPrefixes []string) {
+	pw, err := WatchPodCIDR(clientset)
+	if err != nil {
+		errCh <- err
+		return
+	}
+	defer pw.Stop()
+
+	sw, err := WatchServiceIpAddr(clientset)
+	if err != nil {
+		errCh <- err
+		return
+	}
+	defer sw.Stop()
+
+	for {
+		select {
+		case podSubnet := <-pw.ResultChan():
+			subnet := podSubnet.String()
+			pool, err := getPrefixPool(subnet, additionalPrefixes)
+			if err != nil {
+				logrus.Errorf("Failed to create a prefix pool for '%s' subnet: %v", subnet, err)
+				continue
+			}
+			poolCh <- pool
```

Comment:
`poolCh` is kind of inform channel, because all new coming prefix pools are more actual than unread one. That's why I think we can delete unread prefix pool (if it exists) and write new:
```go
select {
case <-poolCh:
default:
}
poolCh <- pool
```